Twitter API proposal

mr-wildcard opened the pull request:
This is a simple schema for Twitter API. As this API is quite vast, I only implemented three types: `User`, `Tweet`, `Retweet` with few available fields. I also added a `search` query with 3 parameters.

Very simple tests have been provided for Twitter API, but I plan to add tests for `twitter` schema following [graphql-js tests](https://github.com/graphql/graphql-js/blob/master/src/__tests__/starWarsQueryTests.js).

mr-wildcard committed:
update gitignore

for WebStorm, PHPStorm, Node, SublimText.

mr-wildcard committed:
save twit dependency

mr-wildcard committed:
simple twitter object retreived

mr-wildcard committed:
use string id_str to get retweets

mr-wildcard committed:
ass user as UserType field into tweet

mr-wildcard committed:
syntaxx update

mr-wildcard committed:
add user reference to Retweet type

mr-wildcard committed:
get a promise from factory

mr-wildcard committed:
add lodash.get dep

mr-wildcard committed:
add search field + description

mr-wildcard committed:
syntax updates

mr-wildcard committed:
syntax update

mr-wildcard committed:
add retweeted_status ref to retweet type

mr-wildcard committed:
update twitter graphiql url

mr-wildcard committed:
basic test for twitter api

mr-wildcard committed:
add twitter to README

mr-wildcard committed:
clean imports

mr-wildcard committed:
use GraphQLID to represents object id's

Using GraphQLInt would return 'null'.
Previous solution was to type them as GraphQLString. But, as Twitter API
provides both fields 'id' and 'id_str', it seems important to
distinguish them within this schema by picking up the right type.

mr-wildcard committed:
remove id_str field

mr-wildcard committed:
added comment about why using id_str

mr-wildcard committed:
update twitter's playground url

Removed id_str field from graphiql initial query

mr-wildcard committed:
import Twit module from lowercased package name

mr-wildcard committed:
update syntax

clayallsopp commented:
This is awesome! Thank you so much for taking the time to come up with this, will leave some notes, but overall very excited

It looks like Twit is yelling in the Travis builds yelling about missing environment variables; I think you should be able to add some fake ones to [.travis.yml](https://docs.travis-ci.com/user/environment-variables/#Defining-Variables-in-.travis.yml)

clayallsopp commented on .gitignore:
```diff
@@ -1,3 +1,157 @@
-node_modules/
-dist/
+### WebStorm ###
```
where did you pull this .gitignore from? My intuition is to use the [github Node one](https://github.com/github/gitignore/blob/master/Node.gitignore), which covers more Node-specific things

mr-wildcard replied:
Sorry for this brutal change. I pulled it from [gitignore.io](https://www.gitignore.io/) to cover most of the files generated from Webstorm and other IDE's that contributors may use. This gitignore also contains Node's common ignore rules. But I could merge Webstorm and Node specific rules if you wish, tell me.

clayallsopp replied:
gotcha, that makes sense, can you add back in the line about "# Created by https://www.gitignore.io/api/webstorm,node" (which gives a bit more context/guides future users on how to edit it)

clayallsopp commented on apis/twitter.js:
```diff
@@ -0,0 +1,42 @@
+import Twit from 'twit';
+import getObjProperty from 'lodash.get';
```
is there a reason not to use `_.get`?

mr-wildcard replied:
~~I just meant to use a single module from lodash library instead of the entire package. What do you suggest ? :)~~

edit: didn't see that lodash was already installed. this has been fixed.

clayallsopp commented on schemas/graphqlhub.js:
```diff
@@ -9,12 +9,14 @@ import { Schema as HN } from './hn';
 import { Schema as REDDIT } from './reddit';
 import { Schema as KEYVALUE } from './keyvalue';
 import { Schema as GITHUB } from './github';
+import { Schema as Twitter } from './twitter';
```
nit can we keep to the `CAPITALIZE` convention?

mr-wildcard replied:
Sure, i'll udpate my PR.

clayallsopp commented on schemas/twitter.js:
```diff
@@ -0,0 +1,167 @@
+    name: 'TwitterUser',
+    description: 'Twitter user, or as we like to say in France: a \'tweetos\'',
+    fields: () => ({
+        created_at: { type: GraphQLString },
```
nit can we align consecutive colons throughout the file? and put a space between key/colon/value? it's the [style in other schemas](https://github.com/clayallsopp/graphqlhub/blob/master/schemas/hn.js#L257)

filed an issue to integrate ESLint so the tests ask for this automatically https://github.com/clayallsopp/graphqlhub/issues/5

mr-wildcard replied:
I'll try my best to keep consistency with your apis (even if i'm not really comfortable with this alignment style :p)

mr-wildcard replied:
@clayallsopp, even with your example and after reading [ESLint doc about this spacing style](http://eslint.org/docs/rules/key-spacing.html) I'm not completely sure about the modifications I have to provide.
I'm about to update my PR with this kind of spacing: https://gist.github.com/mr-wildcard/bc743991439287b8c7bb
Is this correct or too much ? Please provide an example if it's not ok :)

clayallsopp commented on schemas/twitter.js:
```diff
@@ -0,0 +1,167 @@
+import * as twitter from '../apis/twitter';
+
+import {
+    GraphQLSchema,
```
another style nit - the project generally uses 2 space tabs instead of 4, can we update that?

(basically these style things aim to improve consistency and readability, and in the near future ESLint should enforce them)

mr-wildcard replied:
Sure. I'll indent my files with 2 spaces.

clayallsopp commented on schemas/twitter.js:
```diff
@@ -0,0 +1,167 @@
+                }
+                else {
+
+                    if (isNaN(parseInt(user_id))) {
```
should `_.isNumber` be used?

mr-wildcard replied:
I can add the entire lodash package if you want. (knowing that _.isNumber() simply equals doing `Object.prototype.toString.call(value) ===   '[object Number]'` )

clayallsopp replied:
lodash is already installed: https://github.com/clayallsopp/graphqlhub/blob/master/package.json#L26

mr-wildcard replied:
oooh didn't see that ahah :) Makes perfectly sense to use _.isNumber and _.get then.
I'll update my code + dependencies.

clayallsopp commented on schemas/twitter.js:
```diff
@@ -0,0 +1,167 @@
+    name: 'TwitterAPI',
+    description: 'The Twitter API',
+    fields: {
+        user: {
```
I might prefer splitting this up into two fields - `userById` and `userByScreenName`, each with one `GraphQLNonNull` arg (is there a way in the GraphQL.JS API to have "mutually exclusive" field arguments?)

mr-wildcard replied:
What you think about using a single arg wich is a custom GraphQLScalarType ?
https://facebook.github.io/graphql/#sec-Scalars
http://stackoverflow.com/questions/33819658/dynamic-unique-objects-in-graphql

clayallsopp replied:
:+1: good call

clayallsopp commented on schemas/twitter.js:
```diff
@@ -0,0 +1,167 @@
+                    description: "The number of tweets to return per page, up to a maximum of 100. Defaults to 15. This was formerly the “rpp” parameter in the old Search API."
+                },
+                result_type: {
+                    type: GraphQLString,
```
can this be switched to a `GraphQLEnum`?

mr-wildcard replied:
Yup, this is more appropriate.

clayallsopp commented on schemas/twitter.js:
```diff
@@ -0,0 +1,167 @@
+                },
+                count: {
+                    type: GraphQLInt,
+                    description: "The number of tweets to return per page, up to a maximum of 100. Defaults to 15. This was formerly the “rpp” parameter in the old Search API."
```
can we use the `defaultValue` property instead of putting it in the docs/making it implicit further down in the API?

mr-wildcard replied:
This description has been directly c/c from Twitter API doc. So maybe I'd rather remove limit from doc only as twitter handles this automatically..

clayallsopp commented on README.md:
```diff
@@ -10,6 +10,7 @@ This powers the server behind [GraphQLHub](http://www.graphqlhub.com/). It's bas
 - [Hacker News](schemas/hn.js)
 - [Reddit](schemas/reddit.js)
 - [GitHub](schemas/github.js)
+- [GitHub](schemas/twitter.js)
```
rename to "Twitter"

clayallsopp commented on apis/twitter.js:
```diff
@@ -0,0 +1,42 @@
+    access_token_secret: TWITTER_ACCESS_SECRET
+});
+
+export const getUser = (authKey, authValue) =>  __getPromise('users/show', { [authKey]: authValue });
```
thanks for aligning these :+1:

clayallsopp commented on apis/twitter.js:
```diff
@@ -0,0 +1,42 @@
+    TWITTER_CONSUMER_SECRET
+} = process.env;
+
+const Twitter = new Twit({
```
can we switch this to the [application context](https://github.com/ttezel/twit#twit-api) form? at some point I do want to allow GraphQLHub to handle per-user type of auth tokens, but the pattern thus far is use application-level APIs

mr-wildcard committed:
use appropriate capitalization

mr-wildcard committed:
add dummy env variables for Twit

mr-wildcard committed:
different syntax for travis env

mr-wildcard committed:
add gitignore.io link

mr-wildcard committed:
use main lodash package

mr-wildcard committed:
invert statements

mr-wildcard committed:
indent files with 2 spaces

mr-wildcard committed:
space/alignement for consecutive key/value

mr-wildcard committed:
more key/colon/value spacing

mr-wildcard committed:
update the way GraphQL finds a user

- Use an Enum type for proposing only ‘user_id’ or ‘screen_name’
- Identifier is a Scalar type with checks if the input data is either a
String or an Integer.

mr-wildcard committed:
remove empty under each graphql object instanciation

mr-wildcard committed:
spacing

mr-wildcard committed:
remove bizarre quote.

mr-wildcard committed:
use enum type for search response types

mr-wildcard committed:
add defaultValue to limit

mr-wildcard committed:
change Twit config to app context

mr-wildcard committed:
keep args name the same as graphql ones

mr-wildcard committed:
remove some Travis env variable

mr-wildcard committed:
update twitter playground url

clayallsopp commented:
@mr-wildcard anything else you want to do with this or good-to-merge? LGTM as-is

mr-wildcard commented:
hey @clayallsopp, if it's ok for you then it's ok for me :) let's :shipit:

clayallsopp commented:
It's alive! http://www.graphqlhub.com/playground/twitter
